File: python/graph/max_area_island.py
# You are given an m x n binary matrix grid. An island is a group of 1's (representing land) connected 4-directionally (horizontal or vertical.) You may assume all four edges of the grid are surrounded by water.

# The area of an island is the number of cells with a value 1 in the island.

# Return the maximum area of an island in grid. If there is no island, return 0.


from collections import deque

def maxAreaIsland(grid: list) -> int:
    if not grid:
        return 0
    # get num of rows and cols
    rows, cols = len(grid), len(grid[0])
    # visited coordinates
    visited=set()
    # initial area 
    maxArea = 0
    # search fro neigbors land
    def bfs(r,c):
        q = deque()
        # current island area
        currentArea = 1
        # add visited land
        visited.add((r,c))
        q.append((r,c))
           
        while q:
            # get the land's coordinates
            row,col = q.popleft()
            # directions for the land's neighbors
            directions= [[1,0],[-1,0],[0,1],[0,-1]]
            
            for dr,dc in directions:
                # compute the neighbor's coordinates
                r,c = row + dr, col + dc
                # check for index overflow and unvisited land
                if (r) in range(rows) and (c) in range(cols) and grid[r][c] == 1 and (r ,c) not in visited:
                    # push and visit the new land
                    q.append((r , c ))
                    visited.add((r, c ))
                    # increment the island area
                    currentArea += 1
        return currentArea

    # loop through grid
    for r in range(rows):
        for c in range(cols):
            # find an unvisited land
            if grid[r][c] == 1 and (r,c) not in visited:
                # search down, up, right, left neighbor of this land to find the island area
                maxArea = max(maxArea, bfs(r,c))

    return maxArea

# testing
print(maxAreaIsland(
    [[0,0,1,0,0,0,0,1,0,0,0,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,1,1,0,1,0,0,0,0,0,0,0,0],[0,1,0,0,1,1,0,0,1,0,1,0,0],[0,1,0,0,1,1,0,0,1,1,1,0,0],[0,0,0,0,0,0,0,0,0,0,1,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,0,0,0,0,0,0,1,1,0,0,0,0]]
))
# expected output: 6

File: javascript/graph/max-area-island.js
// ou are given an m x n binary matrix grid. An island is a group of 1's (representing land) connected 4-directionally (horizontal or vertical.) You may assume all four edges of the grid are surrounded by water.

// The area of an island is the number of cells with a value 1 in the island.

// Return the maximum area of an island in grid. If there is no island, return 0.

function maxAreaOfIsland(grid) {
  if (!grid) return 0;
  // get num of rows and cols
  const [rows, cols] = [grid.length, grid[0].length];
  // save visited coordination
  const visited = new Set();
  // numbers of islands
  let maxArea = 0;
  // search for neighbor land
  function bfs(r, c) {
    // simulate a queue
    const q = [];
    // area of the curent island
    let currentArea = 1;
    // add visited land
    visited.add(`${r},${c}`);
    // push the land
    q.push([r, c]);

    while (q.length > 0) {
      // get the land;s coordinates
      const [row, col] = q.shift();
      // directions for the land's neighbors
      const directions = [
        [1, 0],
        [-1, 0],
        [0, 1],
        [0, -1],
      ];

      for (const [dr, dc] of directions) {
        // compute the neighbor's coordinates
        const newR = row + dr;
        const newC = col + dc;
        // check for index overflow and unvisited land
        if (
          newR >= 0 &&
          newR < rows &&
          newC >= 0 &&
          newC < cols &&
          grid[newR][newC] === 1 &&
          !visited.has(`${newR},${newC}`)
        ) {
          // push and visited the new land
          q.push([newR, newC]);
          visited.add(`${newR},${newC}`);
          // increment the island area
          currentArea += 1;
        }
      }
    }

    return currentArea;
  }
  // loop through grid
  for (let row5 = 0; row5 < rows; row5++) {
    for (let col5 = 0; col5 < cols; col5++) {
      // find an unvisited piece of land
      if (grid[row5][col5] === 1 && !visited.has(`${row5},${col5}`)) {
        // search down, up, right, left neighbor of this land to form an island
        maxArea = Math.max(maxArea, bfs(row5, col5));
      }
    }
  }
  return maxArea;
}

// testing
console.log(
  maxAreaOfIsland([
    [0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
    [0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0],
    [0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0],
  ])
);
// expected output: 6